### PR TITLE
fix(login): new login layout

### LINKED ISSUE #1063: Login Page: Refine approach for reporting field-level errors
Currently, field level errors on user name or password will be handled by inserting an error message below the field. This is the standard method for handling form field errors in PatternFly 4. It has been noted that this has the effect of causing elements on the login page to move as error messages are inserted following the field.  This is described in detail in an older design issue opened against PatternFly 3 here: https://github.com/patternfly/patternfly-design/issues/671.

We should explore alternative approaches to  reporting field level errors for the login page to maintain a consistent layout.

Related issues include: #1017 and #910.

### LINKED ISSUE #1069: Login form layout
The remember me checkbox comes after the submit button, which is problematic for screen reader users (i.e. they might never realize it’s there).

During the a11y review meeting, we discussed some options:

- Update the layout so that the checkbox is before/above the submit button
- Update the dom order so that the checkbox is before the submit button, but use css to keep the same layout as we have now.

The first option seemed to be the preferred approach, but we want to run this by design first. If design feels that the change in layout is not preferred, then we can proceed with the second option.

### LINKED ISSUE #1189: Replace images in login footer with FA icons
We don't have permission to use the images in the login page footer, so we should replace those with FA icons and color them to match the brand color.

### LINKED ISSUE #1256: Fix Login component's "Need an account?" text being white
It seems like `--pf-global--Color--100` used to apply to the `p`, but now it doesn't.

Here's an image of the issue (the white text is selected to make it visible):

### PR DESCRIPTION
**Pull request by matthewcarleton**
This closes #1256, closes #1189, closes #1069, Closes #1063

i've added some icon updates too. You'll see new scss vars for dark, light which then need to be updated in the Empty state.

I've added the helper text modifier `pf-m-hidden` to allow us to keep it on the page for spacing but not have it visible to sighted users/screen readers.

The social login links will need a tooltip when bumped to React

### CODE REVIEW

File: src/patternfly/components/Login/docs/code.md [hunk 1]
```diff
@@ -12,11 +12,12 @@
 | `.pf-c-login__main-header .pf-c-dropdown` | `<div>` |  Creates a language selector dropdown in the main header area. |
 | `.pf-c-login__main-body` | `<div>` |  Creates the body of the main area. |
 | `.pf-c-login__main-body .pf-c-form` | `<form>` |  Creates the login form in the main body area. |
+| `.pf-c-login__main-body .pf-c-form .pf-c-form-helper-text.pf-m-error` | `<form>` |  Creates the error text above the form fields, this should always be present on the page with `.pf-m-hidden` applied when it is not active. |
```
Comment (christiemolloy) [inline src/patternfly/components/Login/docs/code.md]: Could we change this to:
`Creates the error text above the form fields. **Required**. When not active, apply `.pf-m-hidden`.

File: src/patternfly/components/Login/docs/code.md [hunk 2]
```diff
@@ -12,11 +12,12 @@
...
 | `.pf-c-login__main-footer` | `<footer>` |  Creates the footer of the main area. |
-| `.pf-c-login__main-footer .pf-c-title` | `<h1>,<h2>,<h3>,<h4>,<h5>,<h6>` |  Creates a title in the main footer area. |
 | `.pf-c-login__main-footer .pf-c-list` | `<ul>` |  Creates a list of links in the main footer area. |
-| `.pf-c-login__main-footer-link` | `<a>` |  Creates links in the main footer area. |
-| `.pf-c-login__main-footer-link-text` | `<span>` |  Creates text in the in the main footer links. |
+| `.pf-c-login__main-footer-links` | `<ul>` |  Creates links in the main footer area. |
```
Comment (christiemolloy) [inline src/patternfly/components/Login/docs/code.md]: Could we change this to:
`Creates an unordered list of links in the main footer`

File: src/patternfly/components/Login/examples/login-invalid-example.hbs [hunk 3]
```diff
@@ -28,71 +27,59 @@
...
-              {{/login-main-footer-link-text}}
-            {{/login-main-footer-link}}
-          {{/login-main-footer-grid-item}}
-          {{#> login-main-footer-grid-item login-main-footer-grid-item--modifier="pf-m-more"}}
-            {{#> button button--attribute='id="login-demo-more"' button--modifier="pf-m-link"}}
-              More...
-            {{/button}}
-          {{/login-main-footer-grid-item}}
-        {{/login-main-footer-grid}}
+        {{#> login-main-footer-links}}
```
Comment (christiemolloy) [inline src/patternfly/components/Login/examples/login-invalid-example.hbs]: The naming of this seems to assume that there will only ever be links in the footer, just checking on this?

File: src/patternfly/components/Login/login.scss [hunk 4]
```diff
@@ -32,44 +32,65 @@
...
+  --pf-c-login__main-header--PaddingRight: var(--pf-global--spacer--xl);
   --pf-c-login__main-header--PaddingBottom: var(--pf-global--spacer--md);
-  --pf-c-login__main-header--PaddingLeft: var(--pf-global--spacer--2xl);
+  --pf-c-login__main-header--PaddingLeft: var(--pf-global--spacer--xl);
+  --pf-c-login__main-header--md--PaddingRight: var(--pf-global--spacer--2xl);
+  --pf-c-login__main-header--md--PaddingLeft: var(--pf-global--spacer--2xl);
   --pf-c-login__main-header--ColumnGap: var(--pf-global--spacer--md);
   --pf-c-login__main-header--RowGap: var(--pf-global--spacer--sm);
 
+  // --pf-c-login__main-header-c-title--md--MarginBottom: var(--pf-global--spacer--lg);
```
Comment (christiemolloy) [inline src/patternfly/components/Login/login.scss]: Remove commented out line

File: src/patternfly/components/Login/login.scss [hunk 5]
```diff
@@ -32,44 +32,65 @@
...
+
+  @media (min-width: $pf-global--breakpoint--md) {
+    --pf-c-login__main-header--PaddingRight: var(--pf-c-login__main-header--md--PaddingRight);
+    --pf-c-login__main-header--PaddingLeft: var(--pf-c-login__main-header--md--PaddingLeft);
+  }
+
   // main body
-  --pf-c-login__main-body--PaddingRight: var(--pf-global--spacer--2xl);
+  // --pf-c-login__main-body--PaddingTop: var(--pf-global--spacer--xl);
```
Comment (christiemolloy) [inline src/patternfly/components/Login/login.scss]: remove commented out line

**Comment by christiemolloy** [comment]
Great job @matthewcarleton, left some comments. One other thing, the input fields are shifting as you click on them, I'm assuming that this is to do with the input component so maybe we can capture an issue there. Not too sure whats causing it, maybe the borders need to be applied ::before

File: src/patternfly/components/Login/examples/login-invalid-example.hbs (hunk 3 above)
Comment (matthewcarleton) [reply src/patternfly/components/Login/examples/login-invalid-example.hbs]: This should always be a list of links. So I think Links is ok but don't feel strongly about it. @mcoker what do you think?

File: src/patternfly/components/Login/login.scss [hunk 6]
```diff
@@ -150,119 +171,107 @@
...
+  .pf-c-form__helper-text {
+    display: flex;
+    align-items: center;
 
-    .pf-c-button {
-      padding-right: var(--pf-c-login__main-body--c-form__group--m-action--c-button--PaddingRight);
-      padding-left: var(--pf-c-login__main-body--c-form__group--m-action--c-button--PaddingLeft);
-      justify-self: stretch;
+    .pf-c-form__helper-text-icon {
+      margin-right: var(--pf-c-login__main-body-c-form-helper-text-icon--MarginRight);
```
Comment (mcoker) [inline src/patternfly/components/Login/login.scss]:
```suggestion
      margin-right: var(--pf-c-login__main-body--c-form__helper-text-icon--MarginRight);
```

File: src/patternfly/components/Login/login.scss [hunk 7]
```diff
@@ -150,119 +171,107 @@
...
+      font-size: var(--pf-c-login__main-body-c-form-helper-text-icon--FontSize);
```
Comment (mcoker) [inline src/patternfly/components/Login/login.scss]:
```suggestion
      font-size: var(--pf-c-login__main-body--c-form__helper-text-icon--FontSize);
```

File: src/patternfly/components/Login/login.scss [hunk 8]
```diff
@@ -32,44 +32,66 @@
...
   --pf-c-login__main-body--PaddingBottom: var(--pf-global--spacer--xl);
-  --pf-c-login__main-body--PaddingLeft: var(--pf-global--spacer--2xl);
-  --pf-c-login__main-body--PaddingTop: var(--pf-global--spacer--xl);
-  --pf-c-login__main-body--c-form__group--m-action--GridGap: var(--pf-global--spacer--lg);
-  --pf-c-login__main-body--c-form__group--m-action--c-button--PaddingRight: var(--pf-global--spacer--xl);
-  --pf-c-login__main-body--c-form__group--m-action--c-button--PaddingLeft: var(--pf-global--spacer--xl);
+  --pf-c-login__main-body--PaddingLeft: var(--pf-global--spacer--xl);
+  --pf-c-login__main-body--md--PaddingRight: var(--pf-global--spacer--2xl);
+  --pf-c-login__main-body--md--PaddingLeft: var(--pf-global--spacer--2xl);
+  --pf-c-login__main-body-c-form-helper-text-icon--FontSize: var(--pf-global--icon--FontSize--md);
```
Comment (mcoker) [inline src/patternfly/components/Login/login.scss]:
```suggestion
  --pf-c-login__main-body--c-form__helper-text-icon--FontSize: var(--pf-global--icon--FontSize--md);
```

File: src/patternfly/components/Login/login.scss [hunk 9]
```diff
@@ -32,44 +32,66 @@
...
+  --pf-c-login__main-body-c-form-helper-text-icon--MarginRight: var(--pf-global--spacer--sm);
```
Comment (mcoker) [inline src/patternfly/components/Login/login.scss]:
```suggestion
  --pf-c-login__main-body--c-form__helper-text-icon--MarginRight: var(--pf-global--spacer--sm);
```

File: src/patternfly/components/Login/login.scss [hunk 10]
```diff
@@ -150,119 +171,107 @@
...
-
-  @media (min-width: $pf-global--breakpoint--md) {
-    .pf-c-login__main-footer-grid-item {
-      .pf-c-button {
-        padding-left: 0;
-      }
-    }
+.pf-c-login__main-footer-band-item {
+  &:first-child {
+    padding-bottom: var(--pf-c-login__main-footer-band-item);
```
Comment (mcoker) [inline src/patternfly/components/Login/login.scss]:
```suggestion
    padding-bottom: var(--pf-c-login__main-footer-band-item--PaddingBottom);
```

However, I might re-write this selector like this to allow for more lines in the future without a CSS change to create this gap.

```
.pf-c-login__main-footer-band-item {
  > * + * {
    padding-top: var(--pf-c-login__main-footer-band-item--PaddingTop);
  }
}
```

File: src/patternfly/components/Login/login.scss [hunk 11]
```diff
@@ -150,119 +171,107 @@
...
-}
 
-.pf-c-login__footer-link {
-  &,
-  &:hover,
-  &.pf-m-hover {
-    color: var(--pf-c-login__footer-link--Color);
+  .pf-c-list {
+    &:not(:only-child) { // paragraph text might not always be included.
+      padding-top: var(--pf-c-login__footer-c-list--PaddingTop);
```
Comment (mcoker) [inline src/patternfly/components/Login/login.scss]:
```suggestion
      padding-top: var(--pf-c-login__footer--c-list--PaddingTop);
```

File: src/patternfly/components/Login/login.scss [hunk 12]
```diff
@@ -32,44 +32,66 @@
...
+  --pf-c-login__main-footer-band--PaddingLeft: var(--pf-global--spacer--md);
+  --pf-c-login__main-footer-band--BackgroundColor: var(--pf-global--BackgroundColor--300);
+  --pf-c-login__main-footer-band--Color: var(--pf-global--Color--300);
+  --pf-c-login__main-footer-band-item: var(--pf-global--spacer--md);
 
   // footer
   --pf-c-login__footer--sm--PaddingLeft: var(--pf-global--spacer--md);
   --pf-c-login__footer--sm--PaddingRight: var(--pf-global--spacer--md);
-  --pf-c-login__footer-link--Color: var(--pf-global--Color--light-100);
+  --pf-c-login__footer-c-list--PaddingTop: var(--pf-global--spacer--md);
```
Comment (mcoker) [inline src/patternfly/components/Login/login.scss]:
```suggestion
  --pf-c-login__footer--c-list--PaddingTop: var(--pf-global--spacer--md);
```

File: src/patternfly/components/Login/login.scss [hunk 13]
```diff
@@ -32,44 +32,66 @@
...
+  --pf-c-login__footer-c-list--xl--PaddingTop: var(--pf-global--spacer--2xl);
+
+  @media (min-width: $pf-global--breakpoint--xl) {
+    --pf-c-login__footer-c-list--PaddingTop: var(--pf-c-login__footer-c-list--xl--PaddingTop);
```
Comment (mcoker) [inline src/patternfly/components/Login/login.scss]:
```suggestion
    --pf-c-login__footer--c-list--PaddingTop: var(--pf-c-login__footer--c-list--xl--PaddingTop);
```

File: src/patternfly/components/Login/login.scss [hunk 14]
```diff
@@ -32,44 +32,66 @@
...
+  --pf-c-login__main-footer-band--BackgroundColor: var(--pf-global--BackgroundColor--300);
+  --pf-c-login__main-footer-band--Color: var(--pf-global--Color--300);
+  --pf-c-login__main-footer-band-item: var(--pf-global--spacer--md);
 
   // footer
   --pf-c-login__footer--sm--PaddingLeft: var(--pf-global--spacer--md);
   --pf-c-login__footer--sm--PaddingRight: var(--pf-global--spacer--md);
-  --pf-c-login__footer-link--Color: var(--pf-global--Color--light-100);
+  --pf-c-login__footer-c-list--PaddingTop: var(--pf-global--spacer--md);
+  --pf-c-login__footer-c-list--xl--PaddingTop: var(--pf-global--spacer--2xl);
```
Comment (mcoker) [inline src/patternfly/components/Login/login.scss]:
```suggestion
  --pf-c-login__footer--c-list--xl--PaddingTop: var(--pf-global--spacer--2xl);
```

**Comment by mcoker** [comment]
https://github.com/patternfly/patternfly-next/blob/deb1aabb66ad05067a1f628eb2d8ffd583132f20/src/patternfly/components/Login/login.scss#L126 should redefine `--pf-c-login__container--MaxWidth`

**Comment by mcoker** [comment]
Hey matt, looks great, left some comments for review.

File: src/patternfly/components/Login/login.scss [hunk 15]
```diff
@@ -32,44 +37,66 @@
...
+  --pf-c-login__main-footer-links--PaddingTop: var(--pf-global--spacer--sm);
+  --pf-c-login__main-footer-links--PaddingRight: var(--pf-global--spacer--3xl);
+  --pf-c-login__main-footer-links--PaddingBottom: var(--pf-global--spacer--xl);
+  --pf-c-login__main-footer-links--PaddingLeft: var(--pf-global--spacer--3xl);
+  --pf-c-login__main-footer-links--MarginBottom: calc(var(--pf-c-login__main-footer-links--MarginBottom) * -1);
+  --pf-c-login__main-footer-links-item--PaddingRight: var(--pf-global--spacer--md);
+  --pf-c-login__main-footer-links-item--PaddingLeft: var(--pf-global--spacer--md);
+  --pf-c-login__main-footer-links-item--MarginBottom: var(--pf-global--spacer--sm);
+  --pf-c-login__main-footer-links-item-link-svg--Fill: var(--pf-global--icon--Color--light);
+  --pf-c-login__main-footer-links-item-link-svg--Width: #{pf-size-prem(25px)};
```
Comment (christiemolloy) [inline src/patternfly/components/Login/login.scss]: are these set to be 25px, or could you use the spacer we have for 24px?

Comment (matthewcarleton) [reply src/patternfly/components/Login/login.scss]: hmm I'm not sure. I pulled these from Andy's PR. I'm not sure we'd want to use a spacer to set the size of these though. The design does call for 24px though and we do have var for icons that size so I'd say it's ok to change it. I don't think anyone will miss 1px :)

**Pull request merged**